Update the MRTK runtime to not allocate a new list on each Update loop.

wiwei opened the pull request:
https://github.com/microsoft/MixedRealityToolkit-Unity/issues/5942

The core MRTK update loop was allocating a new List<> (actually, two, if you look at both Update and LateUpdate). This is because each call to MixedRealityServiceRegistry wasn't being cached. In addition, LINQ adds its own GC overhead (i.e. after re-creating the same identical list, the same list was then sorted again... and again... and again...)

This change avoids that by updating the GetAllServices call to return a cached list of services, which are sorted in priority order. That way the MRTK instead of having to re-copy and resort, can just use the same list on every single frame.

I had also considered keeping the original "sort order" (i.e. random, or actually, insertion order into the registry) but we never formally defined what the ordering of elements was here, and the prime and most important scenario here depends on it being priority order sorted, so we may as well optimize for that use case.

I didn't fully get rid of LINQ (note that the Reverse operation still uses LINQ) - this is okay because it's only in the shutdown path. Alternatively I could have changed the function signature of GetAllServices to be IReadOnlyList, but this would have been a breaking change for not a lot of value.

davidkline-ms commented on Assets/MixedRealityToolkit/Utilities/MixedRealityServiceRegistry.cs:
```diff
@@ -23,6 +25,31 @@ public static class MixedRealityServiceRegistry
             new Dictionary<Type, List<KeyValuePair<IMixedRealityService, IMixedRealityServiceRegistrar>>>();
 
+        /// <summary>
+        /// A cache used to power <seealso cref="GetAllServices(IMixedRealityServiceRegistrar)"/>
```
Nit: power the registrar?

wiwei replied:
Well kinda... the comment as-is is more accurate than just "power the registrar." It powers that specific function, not the entire registrar.

wiwei commented:
Okay, I have to actually update some of the test infra a bit after seeing that unit test failure... this is just adding another test extension service. Doing that now.

wiwei commented:
Alrighty, fixed up the tests. I ended up having to add another test extension service because data providers can't be registered with the registry, and in doing so I added a base test service to reduce all of the copy pasta.

wiwei commented:
/azp run
